tools/cloud_functions/gcs_event_based_ingest/gcs_ocn_bq_ingest/ordering.py

```diff
@@ -0,0 +1,273 @@
+                      "loop until the manual lock is released.")
+                time.sleep(polling_timeout)
+                continue
+        if last_job_done:
+            utils.remove_oldest_backlog_item(gcs_client, bkt, table_prefix)
+            last_job_done = False
+
+        check_backlog_time = time.monotonic()
+        next_backlog_file = utils.get_next_backlog_item(gcs_client, bkt,
+                                                        table_prefix)
+        if not next_backlog_file:
+            backfill_blob.delete(if_generation_match=backfill_blob.generation,
+                                 client=gcs_client)
+            if (check_backlog_time + constants.ENSURE_SUBSCRIBER_SECONDS <
+                    time.monotonic()):
+                print(
+                    "checking if the backlog is still empty for "
+                    f"gs://${bkt.name}/{table_prefix}/_backlog/"
+                    f"There was more than {constants.ENSURE_SUBSCRIBER_SECONDS}"
+                    " seconds between listing items on the backlog and "
+                    f"attempting to delete the {constants.BACKFILL_FILENAME}. "
+                    "This should not happen often but is meant to alleviate a "
+                    "race condition in the event that something caused the "
+                    "delete operation was delayed or had to be retried for a "
+                    "long time.")
+                next_backlog_file = utils.get_next_backlog_item(
+                    gcs_client, bkt, table_prefix)
+                if next_backlog_file:
+                    # The backfill file may have been deleted but the backlog is
+                    # not empty. Re-trigger the backfill subscriber loop by
+                    # dropping a new backfill file.
+                    start_backfill_subscriber_if_not_running(
+                        gcs_client, bkt, table_prefix)
+                    return
+            utils.handle_bq_lock(gcs_client, lock_blob, None)
+            print(f"backlog is empty for gs://{bkt.name}/{table_prefix}. "
+                  "backlog subscriber exiting.")
+            return
+        next_success_file: storage.Blob = bkt.blob(
+            next_backlog_file.name.replace("/_backlog/", "/"))
+        table_ref, batch = utils.gcs_path_to_table_ref_and_batch(
+            next_success_file.name)
+        if not next_success_file.exists(client=gcs_client):
+            raise exceptions.BacklogException(
+                "backlog contains"
+                f"gs://{next_backlog_file.bucket}/{next_backlog_file.name}"
+                "but the corresponding success file does not exist at:"
+                f"gs://{next_success_file.bucket}/{next_success_file.name}")
+        utils.apply(gcs_client, bq_client, next_success_file, lock_blob,
+                    utils.create_job_id(table_ref, batch))
+    # retrigger the subscriber loop by reposting the _BACKFILL file
```

Comment:
Great Question!
Basically if this code never runs, no more BQ jobs will be submitted for this table until this `_BACKFILL` file is replaced.
If other triggers of the function were still able to run the publisher, backlog would accrue but the table would not make progress until this _BACKFILL was replaced.

We can improve this situation by making the `subscriber_monitor` check that the age of the `_BACKFILL` object is not greater than the function timeout. This will mean the system will recover from the abandoned `_BACKFILL` the next time a file is added to the backlog. (I'll prepare this in my next commit)

a temporary network "hiccup" shouldn't matter much as client library will retry requests and we build in some healthy buffer before the function actually times out however a true network failure might cause these retries to be futile.

If cloud functions platform goes down in the region, we won't really be able to recover w/ in this cloud function. It could be possible to run this function in multiple regions, in this case one should silence the logs about duplicate notifications as this would be "by design" as you'd expect a race between the functions running in each region to claim the work for a particular event.